### ISSUE TITLE
Linking into an iOS app generates warnings in 2.0.3 and 2.0.4

### ISSUE THREAD

**Issue by dhoerl** [libjpeg-turbo/libjpeg-turbo#406]
**Have you searched the existing issues (both open and closed) in the libjpeg-turbo issue tracker to ensure that this bug report is not a duplicate?**

Yes.

**Does this bug report describe one of the [two known and unsolvable issues with the JPEG format](https://libjpeg-turbo.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf)?**

No.

**Clear and concise description of the bug:**

When I link 2.0.2 into my iOS app to test in the Simulator, it links just fine. 2.0.3 & 2.0.4 both generate these warnings (Xcode 11.3 and 11.4):

ld: warning: building for iOS Simulator, but linking in object file (/Volumes/Data/git/PhotoScrollerSwiftPackage/Libraries/libturbojpeg.a(jdapimin.c.o)) built for macOS

However, I find the App still runs - I have a unit test that uses the library and it still passes.

**Steps to reproduce the bug (using *only* libjpeg-turbo):**

Install the 2.0.4 (or 2.0.3) DMG in /opt. Copy the libturbojpeg.a from the lib directory (or just link to the archive in this location). Build generates warnings when targeted to the Simulator.

Run uninstall script

Install the 2.0.2 DMG in /opt. Copy the libturbojpeg.a from the lib directory (or just link to the archive in this location). Now app builds fine, no warnings.

**Image(s) needed in order to reproduce the bug (if applicable):**

**Expected behavior:**

**Observed behavior:**

**Platform(s) (compiler version, operating system version, CPU) on which the bug was observed:**

Xcode 11, Mac Catalina

**libjpeg-turbo release(s), commit(s), or branch(es) in which the bug was observed (always test the tip of the master branch or the latest [stable pre-release](https://libjpeg-turbo.org/DeveloperInfo/PreReleases) to verify that the bug hasn't already been fixed):**

**If the bug is a regression, the specific commit that introduced the regression (use `git bisect` to determine this):**

**Additional information:**

**Comment by dhoerl** [comment]
OK - found a thread on this in the Apple dev forums:

https://forums.developer.apple.com/message/364420#364420

I believe that there needs to be two libraries - one for iOS and one for Mac. They could both be in the same directory as now, maybe have the iOS library named "libturbojpegIOS.a"?

Is the code to do the DMG here? (I'll go poke around.) If I find it will see if I can get something working...

**Comment by dcommander** [comment]
You're going to have to provide more details.  Obviously it isn't as simple as just using Xcode, because I use Xcode to build libjpeg-turbo, and the example applications link just fine with libturbojpeg.a.  libturbojpeg.a contains the required architectures for iOS:

```
> lipo -info /opt/libjpeg-turbo/lib/libturbojpeg.a 
Architectures in the fat file: /opt/libjpeg-turbo/lib/libturbojpeg.a are: i386 armv7 armv7s x86_64 arm64 
```

The forum post is irrelevant, because libjpeg-turbo 2.0.2 obviously does not have a separate iOS library, and yet you are able to use it successfully.  I do not know of any differences in how 2.0.2 and 2.0.3 were built.  I'm pretty sure I used the same version of Xcode to build both.

**Comment by dcommander** [comment]
If you could supply a simple Xcode project that demonstrates the problem, that would be very helpful.

**Comment by dhoerl** [comment]
Project is here: https://www.dropbox.com/s/3w2jnfjq2xshhl6/LTJTest.zip?dl=0

There are two libraries in it 202 and 204. Right Link uses 202 and there are now errors. Drag it out, drag in 204, type CMD-B (with a simulator selected - I'm using an iPad one). 

With 2.0.4 you see the errors.

I suspect that clang changed between 2.0.2 and 2.0.3. Looking at that forum link others had the same issue late last summer. A guy from Apple opined that the MacOS slices have a different signature than the iOS ones (for the Intel slices), and that it would NOT be possible to build a single library with both iOS and Mac.

Trust me  - I looked at the architectures too before writing this - I was looking for something different - but never found it.

**Comment by dcommander** [comment]
I don't see how the Clang version could have changed.  As you can see from the build scripts (https://github.com/libjpeg-turbo/buildscripts), the use of Xcode 7.2 is hard-coded into the iOS builds and has been since 1.5.1, when bitcode was introduced into the iOS binaries.

**Comment by dhoerl** [comment]
Well, all I'm saying is when I link with 202, it works. When I link with 204, the build log is filled with those warnings (in the example I uploaded its only two lines, but in a real app its around 60). The Forum link shows many other open source providers dealing with the same issue. It appears you need to provide a specialized target for iOS: "-target x86_64-apple-ios13.0-macabi" - for example

**Comment by dcommander** [comment]
Looking at the official build logs, the only major difference I see is that I upgraded my build system from 10.11 "El Capitan" to 10.12 "Sierra" between 2.0.2 and 2.0.3.  That had zero effect on the iOS builds, but I did start seeing build warnings in the x86-64 and i386 builds after that.  Those warnings seemed to be due to an incompatibility between GCC 5 in MacPorts and macOS 10.12 and later.  I stopped seeing the warnings when I switched to GCC 6, but that change is only currently active in the libjpeg-turbo 2.1 pre-release builds.

Unfortunately, Apple's forced obsolescence BS has made it impossible for me to reproduce this problem.  As you probably know, best practices in multi-platform software development include building on the oldest platform you need to support, so my macOS build machine is always two generations behind.  I also keep it back-dated because some of the 3rd party content creation software I use is notoriously behind the curve and never supports the latest macOS.  But of course Apple's own software, including Xcode, is the opposite-- usually requiring you to update your O/S just to get a bug fix.  As a result, I cannot run Xcode 11, so I cannot use the iOS 13 SDK.

Action item:
* Please test the libjpeg-turbo 2.1 evolving pre-release (https://libjpeg-turbo.org/DeveloperInfo/PreReleases) and see if it fixes the problem.  If it does, then it's easy enough for me to rebuild the 2.0.4 binaries using GCC 6.

**Comment by dhoerl** [comment]
I completely understand your reasoning, and if I can help (as I have in the past), willing to do that.

Will test pre-release shortly. Also will see if there is a linker flag to disable the warning.

This is just an annoyance - the app runs fine in the simulator. I dislike warning since they often mask real problems that creep in. But given everything else, its a really minor problem.

**Comment by dcommander** [comment]
OK, let me know how it goes.  In the future, please be mindful of semantics.  You used the terms "broke" and "errors" to describe this problem.  Neither of those words is indicative of a minor problem.

**Comment by dhoerl** [comment]
I had to edit the original post. When I ran it first time, the Simulator froze - I assumed it was the library. Later on wasn't sure I did it right, had to fuss with Xcode and Simulator for a while to get even 202 to run. Then tried again, found 204 would run - so I edited (poorly) the original post. But understand your comment.

The new 2.0.80 is way worse. I cannot even build it now:

error: Building for iOS Simulator, but the linked library '208libturbojpeg.a' was built for macOS + iOS. (in target 'LTJTest' from project 'LTJTest')

Lipo shows only that this 2.0.80 prerelease has two less arch - nothing else:

 lipo -detailed_info 208libturbojpeg.a
Fat header in: 208libturbojpeg.a
fat_magic 0xcafebabe
nfat_arch 3
architecture i386
    cputype CPU_TYPE_I386
    cpusubtype CPU_SUBTYPE_I386_ALL
    offset 68
    size 990528
    align 2^2 (4)
architecture x86_64
    cputype CPU_TYPE_X86_64
    cpusubtype CPU_SUBTYPE_X86_64_ALL
    offset 990600
    size 979256
    align 2^3 (8)
architecture arm64
    cputype CPU_TYPE_ARM64
    cpusubtype CPU_SUBTYPE_ARM64_ALL
    offset 1969856
    size 1465064
    align 2^3 (8)

**Comment by dcommander** [comment]
To be clear, are you trying to build the app using the x86-64 architecture for the simulator, or are you using the ARM architecture? I think the crux of the issue may be that the i386 and x86-64 architectures don’t have bitcode and aren’t built for iOS at all, but that doesn’t explain why 2.0.2 worked.

**Comment by dhoerl** [comment]
I'm doing the same thing I have for like 10 years. I have an open source project that uses libturbojpeg to process data in chunks as it downloads to an iOS device. It writes tiles to the file system, so it can display huge images.

Apps that use it will be tested on the simulator, then deployed to the field as iOS apps. So I need both ARM and INTEL slices. Apple keeps telling devs that they should not bundle ARM and INTEL into one archive, but it has always seems to work. If you have two archives, one for the Simulator and one for Devices, its a PITA to add them - you need to add different link likes in Xcode, and its easy to get it wrong. Perhaps if I did that the warnings would go away.

Frankly, now I'm more worried about future releases not even linking. I have an old open source project to build libturbojpeg with CLANG, which as we discussed at length about 4 years ago doesn't perform as well as gcc. But its a backup plan anyway.

**Comment by dcommander** [comment]
You didn't answer my question.  Are you building an ARM app for iOS deployment or an x86-64 app for the simulator when you see the warnings?  Or are the warnings seen in both cases?

**Comment by dhoerl** [comment]
The app builds fine when a "Generic iOS Target" is selected - no warnings - for 202, 204, and 2.0.80. All the issues are when building for the Simulator.

Interestingly enough, that LTJTest App was created in Xcode 11.4-beta (what I'm mostly using). The previous warnings are now hard errors (Simulator only):

error: Building for iOS Simulator, but the linked library '208libturbojpeg.a' was built for macOS + iOS. (in target 'LTJTest' from project 'LTJTest')

I'm poking around to see what has changed (if anything obvious). When I build with 2.0.4 in Xcode 11, using a project created in Xcode 11.3, just warnings. Weird.

I'm going to post a question on the Apple forums, will update here with any responses.

**Comment by dcommander** [comment]
I'm beginning to develop a better understanding of the problem.  It seems that the warnings I observed and the warnings you observed are tangentially related.

* For whatever reason, after I upgraded my build machine from El Cap to Sierra, this caused MacPorts GCC 5 to start embedding the `LC_VERSION_MIN_MACOSX` load command into the object files it generates.  This may have been triggered by updating MacPorts GCC 5 to the latest version, which I would have done after the O/S upgrade.
* For whatever reason, MacPorts GCC 5 does not properly respect the `-mmacosx-min-version` compiler argument, so it sets `LC_VERSION_MIN_MACOSX` to the operating system version rather than the version specified by `-mmacosx-min-version`.  This was the source of the warnings I started observing in the official builds.
* MacPorts GCC 6 apparently fixed that problem, which is why the warnings in the official builds went away.
* Whenever `LC_VERSION_MIN_MACOSX` is set to any value, it triggers warnings when attempting to link an iOS simulator app.  Those warnings do not occur with libjpeg-turbo <= 2.0.2 because those binaries do not contain the `LC_VERSION_MIN_MACOSX` load command.

If there is a way to prevent GCC from embedding `LC_VERSION_MIN_MACOSX`, I'm not sure how, and it seems that the Mac App Store now requires that load command anyhow.

Since it also seems impossible to create a fat binary with both a simulator and macOS build, that leaves us with limited options:
1. Ignore the warnings whenever building for the iOS simulator.  (But that seems to no longer be an option with Xcode 11.4 beta, if I understand you correctly.)
2. Create separate libraries for the iOS simulator.  (But that would require disruptive changes to the build/packaging system.  That system is designed to generate fat binaries with multiple targets/architectures, not separately-installed binaries for different targets/architectures.)
3. Create a separate libjpeg-turbo package for the iOS simulator.  This can be done by modifying only the official build scripts.
4. Require that iOS simulator applications use a custom build of libjpeg-turbo.  It is straightforward to generate such a build.  Example:
    ```
    wget https://sourceforge.net/projects/libjpeg-turbo/files/2.0.4/libjpeg-turbo-2.0.4.tar.gz
    tar xfz libjpeg-turbo-2.0.4.tar.gz
    cd libjpeg-turbo-2.0.4
    mkdir build
    cd build
    cmake -G"Unix Makefiles" -DCMAKE_C_FLAGS='-miphoneos-version-min=8.0' -UCMAKE_OSX_DEPLOYMENT_TARGET -DCMAKE_OSX_SYSROOT=`xcode-select -p`/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk -DCMAKE_POSITION_INDEPENDENT_CODE=1 -DREQUIRE_SIMD=1 -DCMAKE_ASM_NASM_COMPILER=/opt/local/bin/yasm ..
    ```

(1) or (4) would be my preferred solutions.  Generating separate binaries just for the simulator seems like a rabbit hole.

**Comment by dhoerl** [comment]
Using the #4 above, and breaking the long line into pieces (adding reference to src dir too):
```
SRCDIR=$PWD
cd /tmp
cmake -G"Unix Makefiles" -DCMAKE_C_FLAGS='-miphoneos-version-min=8.0' \ 
	-UCMAKE_OSX_DEPLOYMENT_TARGET \
	-DCMAKE_OSX_SYSROOT=`xcode-select -p `/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk \
	-DCMAKE_POSITION_INDEPENDENT_CODE=1 \
	-DREQUIRE_SIMD=1 \
	-DCMAKE_ASM_NASM_COMPILER=/usr/local/Cellar/yasm "$SRCDIR"

```
I get this error:

 -UCMAKE_OSX_DEPLOYMENT_TARGET: not found

**Comment by dhoerl** [comment]
You are correct on 1) - at least in this beta, the error is hard. I'm fine with 4) - I had a GitHub project to build latest or any previous version o the library, I'm going to get it back into shape. I was also able to get 4) to build. I'll test the lipo'd archive Monday. Thanks!

**Comment by dhoerl** [comment]
One final thought - the solution you offer today with the pre-packaged binaries is mixed: meaning, you have 3 iOS slices and one MacOS slices. Why not just make the pre-built archive MacOS only, and let iOS devs build what they need (probably just 64 bit ios+simulator)? Just a thought.

**Comment by dcommander** [comment]
Yes, because I didn't know until now that the x86[-64] binaries could not be used with the iOS simulator.  That is a new development.  The current slicing scheme was one you recommended some years ago.

I am opposed to feature regression, so I think it makes the most sense to provide a separate package containing all of the iOS slices (ARM and x86 simulator.)  I will think about what makes the most sense in terms of deployment, but most likely, this will land in 2.1.

**Comment by dcommander** [comment]
Unfortunately, the new packaging scheme required modifications to the libjpeg-turbo build system, so it couldn't be deployed against the existing 2.0.4 source code.  It has been implemented in the dev branch and is now available for testing in the 2.1 evolving pre-release builds: https://libjpeg-turbo.org/DeveloperInfo/PreReleases.  In those builds, the regular libjpeg-turbo DMG contains only i386 and x86-64 macOS slices.  The new libjpeg-turbo-ios DMG, which is installed into a different directory (**/opt/libjpeg-turbo-ios**), contains i386, x86-64 and ARMv8 iOS slices (ARMv7 and ARMv7s slices are no longer included.)  https://libjpeg-turbo.org/Documentation/OfficialBinaries has been updated accordingly.  Please test and let me know if it works.
